external-crates/move/crates/move-bytecode-verifier/src/check_duplication.rs

```diff
@@ -239,20 +252,111 @@ impl<'a> DuplicationChecker<'a> {
             ));
         }
         // Check that each struct handle in self module is implemented (has a declaration)
-        let implemented_struct_handles: HashSet<StructHandleIndex> = self
+        let implemented_datatype_handles: HashSet<DatatypeHandleIndex> = self
             .module
             .struct_defs()
             .iter()
             .map(|x| x.struct_handle)
+            .chain(self.module.enum_defs().iter().map(|x| x.enum_handle))
+            .collect();
+        if let Some(idx) = (0..self.module.datatype_handles().len()).position(|x| {
+            let y = DatatypeHandleIndex::new(x as u16);
+            self.module.datatype_handle_at(y).module == self.module.self_handle_idx()
+                && !implemented_datatype_handles.contains(&y)
+        }) {
+            return Err(verification_error(
+                StatusCode::UNIMPLEMENTED_HANDLE,
+                IndexKind::DatatypeHandle,
+                idx as TableIndex,
+            ));
+        }
+        Ok(implemented_datatype_handles)
+    }
+
+    fn check_enum_definitions(&self) -> PartialVMResult<()> {
+        // EnumDefinition - contained DatatypeHandle defines uniqueness
+        // NB: We check uniqueness across both enum and struct handles at this point to make sure
+        // data definitions are not duplicated across struct and enums.
+        if let Some(idx) = Self::first_duplicate_element(
+            self.module
+                .struct_defs()
+                .iter()
+                .map(|x| x.struct_handle)
+                .chain(self.module.enum_defs().iter().map(|x| x.enum_handle)),
+        ) {
+            return Err(verification_error(
+                StatusCode::DUPLICATE_ELEMENT,
+                IndexKind::EnumDefinition,
+                idx,
+            ));
+        }
+        // Variant names in enums must be unique
+        // Field names in each variant must be unique
+        for (enum_idx, enum_def) in self.module.enum_defs().iter().enumerate() {
+            let variants = &enum_def.variants;
+            if variants.is_empty() {
+                return Err(verification_error(
+                    StatusCode::ZERO_SIZED_ENUM,
+                    IndexKind::EnumDefinition,
+                    enum_idx as TableIndex,
+                ));
+            }
+            if let Some(idx) =
+                Self::first_duplicate_element(variants.iter().map(|x| x.variant_name))
+            {
+                return Err(verification_error(
+                    StatusCode::DUPLICATE_ELEMENT,
+                    IndexKind::EnumDefinition,
+                    enum_idx as TableIndex,
+                )
+                .at_index(IndexKind::VariantTag, idx as TableIndex));
+            }
+
+            // NB: we allow zero-sized variants: since we require non-empty enums we always have a
```

Comment:
Basically, we allow variants to not have any fields so if you have something like
```
public enum Foo {
   Bar
}
```
We won't add a dummy bool field to `Bar` (a la structs). This is because we require at least one variant per enum (so we get the non-empty property from that).

Reply:
Ah I see, empty variants but non empty enums.

Side note, having empty enums can be kind of helpful, though I understand it might push the complexity a bit. Definitely don't add it yet, but its something we should think about it